File: frontend/ts/picture/figure/Ellipse.ts
import {Point} from "./Point.js";
import {TrivialFigure} from "./TrivialFigure.js";

export class Ellipse extends TrivialFigure {
    constructor(
        color: string,
        public readonly center: Point,
        public readonly radius: Point
    ) {
        super("ellipse", color)
    }

    draw(ctx: CanvasRenderingContext2D): void {
        ctx.fillStyle = this.color;
        ctx.beginPath();
        ctx.ellipse(
            this.center.x, this.center.y,
            this.radius.x, this.radius.y,
            0, 0, 2 * Math.PI
        );
        ctx.fill();
    }
}

export namespace Ellipse {
    export function fromJson(json: Map<string, any>): Ellipse {
        return new Ellipse(
            json.get("color") as string,
            Point.fromJson(json.get('center') as Map<string, any>),
            Point.fromJson(json.get('radius') as Map<string, any>)
        )
    }
}

File: backend/src/main/java/ru/vityaman/itmo/web/lab/servlets/backend/model/figure/Ellipse.java
package ru.vityaman.itmo.web.lab.servlets.backend.model.figure;

import lombok.Value;
import ru.vityaman.itmo.web.lab.servlets.backend.model.Hashable;

import static java.lang.Math.pow;

@Value
public final class Ellipse extends TrivialFigure {
    private final Point center;
    private final Radius radius;

    public Ellipse(Color color, Point center, Radius radius) {
        super("ellipse", color);
        this.center = center;
        this.radius = radius;
    }

    public static Ellipse circle(Color color, Point center, long radius) {
        return new Ellipse(color, center, new Radius(radius, radius));
    }

    @Override
    public boolean contains(Point point) {
        long dx = point.x() - center.x();
        long dy = point.y() - center.y();
        return pow(dx, 2) / pow(radius.x(), 2)
            + pow(dy, 2) / pow(radius.y(), 2) <= 1;
    }

    @Override
    public String hash() {
        return Hashable.hash(
            String.format(
                "ellipse:%s:%s:%s:%s",
                center.x(), center.y(),
                radius.x, radius.y
            )
        );
    }

    @Value
    public static final class Radius {
        private final long x;
        private final long y;
    }
}

